android/PhysicalWeb/app/src/main/java/org/physical_web/physicalweb/NearbyBeaconsFragment.java

```diff
@@ -337,16 +346,57 @@ private void emptyGroupIdQueue() {
 
     for (String groupId : mGroupIdQueue) {
       Log.d(TAG, "groupid " + groupId);
-      pwPairs.add(Utils.getTopRankedPwPairByGroupId(mPwCollection, groupId));
+      pwPairs.add(Utils.getTopRankedPwPairByGroupId(mPwCollection, groupId, mFavorites));
     }
-    Collections.sort(pwPairs, Utils.newDistanceComparator());
+    Collections.sort(pwPairs, new Utils.FavoritesComparator(mFavorites));
     for (PwPair pwPair : pwPairs) {
       mNearbyDeviceAdapter.addItem(pwPair);
     }
     mGroupIdQueue.clear();
     mNearbyDeviceAdapter.notifyDataSetChanged();
   }
 
+  public boolean isFavorite(String siteUrl) {
+    return mFavorites.contains(siteUrl);
+  }
+
+  public void flipFavorite(String siteUrl) {
+    if (isFavorite(siteUrl)) {
+      mFavorites.remove(siteUrl);
+      return;
+    }
+    mFavorites.add(siteUrl);
+  }
+
+  private void saveFavorites() {
+    // Write the PW Collection
+    String preferencesKey = getString(R.string.discovery_service_prefs_key);
+    SharedPreferences prefs = getActivity().getSharedPreferences(
+        preferencesKey, Context.MODE_PRIVATE);
+    SharedPreferences.Editor editor = prefs.edit();
+    try {
+      editor.putString(Utils.FAVORITES_KEY, (new JSONArray(mFavorites.toArray())).toString());
```

Comment:
You could use editor.putStringSet to store the favorites list rather than serializing the JSON array